File: app/src/main/java/com/icerrate/popularmovies/data/source/MovieRepository.java
package com.icerrate.popularmovies.data.source;

import com.icerrate.popularmovies.data.model.Movie;
import com.icerrate.popularmovies.data.model.PaginatedResponse;
import com.icerrate.popularmovies.data.model.Review;
import com.icerrate.popularmovies.data.model.Trailer;
import com.icerrate.popularmovies.data.model.TrailersResponse;
import com.icerrate.popularmovies.view.common.BaseCallback;

import java.util.ArrayList;

import static com.google.gson.internal.$Gson$Preconditions.checkNotNull;

/**
 * Created by Ivan Cerrate.
 */

public class MovieRepository implements MovieDataSource {

    private MovieDataSource localDataSource;

    private MovieDataSource remoteDataSource;

    private static MovieRepository INSTANCE = null;

    public MovieRepository(MovieDataSource localDataSource, MovieDataSource remoteDataSource) {
        this.localDataSource = checkNotNull(localDataSource);
        this.remoteDataSource = checkNotNull(remoteDataSource);
    }

    public static MovieRepository getInstance(MovieDataSource localDataSource, MovieDataSource remoteDataSource) {
        if (INSTANCE == null) {
            INSTANCE = new MovieRepository(localDataSource, remoteDataSource);
        }
        return INSTANCE;
    }

    @Override
    public void getPopularMovies(Integer page, BaseCallback<PaginatedResponse<Movie>> callback) {
        remoteDataSource.getPopularMovies(page, callback);
    }

    @Override
    public void getTopRatedMovies(Integer page, BaseCallback<PaginatedResponse<Movie>> callback) {
        remoteDataSource.getTopRatedMovies(page, callback);
    }

    @Override
    public void getMovieTrailers(Integer movieId, BaseCallback<TrailersResponse<Trailer>> callback) {
        remoteDataSource.getMovieTrailers(movieId, callback);
    }

    @Override
    public void getMovieReviews(Integer movieId, BaseCallback<PaginatedResponse<Review>> callback) {
        remoteDataSource.getMovieReviews(movieId, callback);
    }

    @Override
    public void getFavoriteMovies(BaseCallback<ArrayList<Movie>> callback) {
        localDataSource.getFavoriteMovies(callback);
    }

    @Override
    public void isFavoriteMovie(int movieId, BaseCallback<Boolean> callback) {
        localDataSource.isFavoriteMovie(movieId, callback);
    }

    @Override
    public void addFavoriteMovie(Movie movie, BaseCallback<Void> callback) {
        localDataSource.addFavoriteMovie(movie, callback);
    }

    @Override
    public void removeFavoriteMovie(int movieId, BaseCallback<Void> callback) {
        localDataSource.removeFavoriteMovie(movieId, callback);
    }
}

File: app/src/main/java/com/icerrate/popularmovies/data/source/MovieRepository.kt
package com.icerrate.popularmovies.data.source

import com.icerrate.popularmovies.data.model.Movie
import com.icerrate.popularmovies.data.model.PaginatedResponse
import com.icerrate.popularmovies.data.model.Resource
import com.icerrate.popularmovies.data.model.Review
import com.icerrate.popularmovies.data.model.Trailer
import com.icerrate.popularmovies.data.model.TrailersResponse

/**
 * @author Ivan Cerrate.
 */
class MovieRepository private constructor(
    private val localDataSource: MovieDataSource,
    private val remoteDataSource: MovieDataSource
) : MovieDataSource {

    companion object {
        @Volatile
        private var INSTANCE: MovieRepository? = null

        fun getInstance(
            localDataSource: MovieDataSource,
            remoteDataSource: MovieDataSource
        ): MovieRepository {
            return INSTANCE ?: synchronized(this) {
                INSTANCE ?: MovieRepository(localDataSource, remoteDataSource).also { INSTANCE = it }
            }
        }
    }

    override suspend fun getPopularMovies(page: Int?): Resource<PaginatedResponse<Movie>> =
        remoteDataSource.getPopularMovies(page)

    override suspend fun getTopRatedMovies(page: Int?): Resource<PaginatedResponse<Movie>> =
        remoteDataSource.getTopRatedMovies(page)

    override suspend fun searchMovies(
        query: String,
        page: Int?
    ): Resource<PaginatedResponse<Movie>> =
        remoteDataSource.searchMovies(query, page)

    override suspend fun getMovieTrailers(movieId: Int?): Resource<TrailersResponse<Trailer>> =
        remoteDataSource.getMovieTrailers(movieId)

    override suspend fun getMovieReviews(movieId: Int?): Resource<PaginatedResponse<Review>> =
        remoteDataSource.getMovieReviews(movieId)

    override suspend fun getFavoriteMovies(): Resource<ArrayList<Movie>> =
        localDataSource.getFavoriteMovies()

    override suspend fun isFavoriteMovie(movieId: Int): Resource<Boolean> =
        localDataSource.isFavoriteMovie(movieId)

    override suspend fun addFavoriteMovie(movie: Movie): Resource<Unit> =
        localDataSource.addFavoriteMovie(movie)

    override suspend fun removeFavoriteMovie(movieId: Int): Resource<Unit> =
        localDataSource.removeFavoriteMovie(movieId)
}